### DISCUSSION TITLE
WPF UI更新不流畅

### DISCUSSION THREAD

**Discussion by zuogithub** [BYJRK/DotNet-Discussions#125] [WPF 开发]
**创建线程对绑定的数据连续更新，发现UI上刷新的值时不时会卡顿一下，出现这个情况的原因是什么？怎么解决？**
比如视频中值为228，247，264，290的时候

https://github.com/BYJRK/DotNet-Discussions/assets/66720317/ba5de60b-4ed0-4de7-bc62-ad4bf3c62a94

xaml

    <Window x:Class="WpfApp1.MainWindow"

        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"

        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"

        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"

        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"

        xmlns:local="clr-namespace:WpfApp1"

        mc:Ignorable="d"

        Title="MainWindow" Height="450" Width="800">

    <Window.DataContext>

        <local:MainWindowViewModel/>

    </Window.DataContext>

    <Grid>

        <TextBox HorizontalAlignment="Left" Margin="31,24,0,0" TextWrapping="Wrap" Text="{Binding Cnt}" VerticalAlignment="Top" Width="218" Height="24"/>

        <TextBox HorizontalAlignment="Left" Margin="31,76,0,0" TextWrapping="Wrap" Text="{Binding Dt}" VerticalAlignment="Top" Width="218" Height="24"/>

    </Grid>

    </Window>

C#

     using CommunityToolkit.Mvvm.ComponentModel;
    using System;
    using System.Threading;
    using System.Threading.Tasks;
    namespace WpfApp1
    {

    internal partial class MainWindowViewModel: ObservableObject
    {
        public MainWindowViewModel()
        {
            last = DateTime.Now;

            Task readAxiSatatusTask = Task.Factory.StartNew(() =>

            {

                while (true)

                {

                    Thread.Sleep(100);

                    if (Cnt < 100000)

                        Cnt += 1;

                    else

                        Cnt = 0;

                    Dt = (DateTime.Now - last).TotalMilliseconds;

                    last = DateTime.Now;

                }

            }, TaskCreationOptions.LongRunning);

        }
        [ObservableProperty]
        private int cnt;
        private DateTime last;
        [ObservableProperty]
        private double dt=0;
    }
    }

**Comment by CodingOctocat** [comment]
把 readAxiSatatusTask  逻辑移动到 Loaded 方法，`Thread.Sleep(100);` 改成 `await Task.Delay(100);`

```CS
private Task ReadAxiSatatusAsync()
{
    while (true)
    {

        await Task.Delay(100);

        if (Cnt < 100000)

            Cnt += 1;

        else

            Cnt = 0;

        Dt = (DateTime.Now - last).TotalMilliseconds;

        last = DateTime.Now;
    }
}

// MvvmToolkit 将生成 LoadedCommand 命令（* 忽略 Async 后缀）。
[RelayCommand]
private async void LoadedAsync()
{
    await ReadAxiSatatusAsync();
}
```

借助 https://github.com/microsoft/XamlBehaviorsWpf 将 LoadedCommand 命令绑定到 View。

以下代码放到 MainWindow 根节点：

```XAML
<i:Interaction.Triggers>
    <i:EventTrigger EventName="Loaded">
        <i:InvokeCommandAction Command="{Binding LoadedCommand}" />
    </i:EventTrigger>
</i:Interaction.Triggers>
```

**Reply by zuogithub** [reply]
感谢，试了确实有效果，能麻烦讲一下原因吗？初学WPF，有点不太明白。

**Reply by CodingOctocat** [reply]
@zuogithub 因为 `Thread.Sleep;` 会挂起/阻塞线程，所以会卡，你要换成异步延迟：`await Task.Delay`，但是构造函数又不支持异步(async 修饰符)，所以将逻辑放到 Loaded。

**Reply by zuogithub** [reply]
@CodingOctocat  readAxiSatatusTask与UI线程是同一个线程吗?还是说readAxiSatatusTask不一定在哪个线程，在UI线程的时候就会阻塞UI？

**Reply by CodingOctocat** [reply]
@zuogithub 应该不在同一个线程吧。

**Reply by zuogithub** [reply]
@CodingOctocat 大概明白了，谢谢

**Reply by CodingOctocat** [reply]
@zuogithub 在不在 UI 线程你可以试着在那段代码里去修改一个 UI 属性，比如 MainWindow 放一个 `TextBlock`，然后随便设置下它的 Text 属性值，如果报错了说明不在 UI 线程。或者使用 `Dispatcher.CurrentDispatcher.CheckAccess()` 判断。

**Reply by zuogithub** [reply]
@CodingOctocat  
**通过给`Task.Delay`添加断点，多次中断，发现我原来的方式会中断在不同的线程**
<img width="154" alt="0ace8bdd2f899f8137dfefaec3eeda0" src="https://github.com/BYJRK/DotNet-Discussions/assets/66720317/a3a4d137-432c-44a1-b40b-154b5cae04a6">
<img width="155" alt="微信截图_20230803174331" src="https://github.com/BYJRK/DotNet-Discussions/assets/66720317/e18f2129-885d-413f-bb2a-f52a97a62c1b">
<img width="152" alt="54b7741a0366f733101ead8c7bc1e4a" src="https://github.com/BYJRK/DotNet-Discussions/assets/66720317/e4d4c5a7-7471-4d2f-b3ea-0cc8a6a8b37f">

**按照您提供的方式每次都中断在主线程**
<img width="156" alt="微信截图_20230803173723" src="https://github.com/BYJRK/DotNet-Discussions/assets/66720317/9c80ba07-85e0-4318-a20b-ebf21e352936">

**按照您的思路，其实直接在MainWindowViewModel构造函数里面调用异步方法`ReadAxiSatatusAsync()`也是可以的。**

    public MainWindowViewModel()
        {
            ReadAxiSatatusAsync();
        }
    private async Task ReadAxiSatatusAsync()
        {
            while (true)
            {

                await Task.Delay(1000);

                if (Cnt < 100000)

                    Cnt += 1;

                else

                    Cnt = 0;

                Dt = (DateTime.Now - last).TotalMilliseconds;

                last = DateTime.Now;
            }
        }

**Comment by CodingOctocat** [comment]
你也可以试试仅仅把 `Thread.Sleep` 换成 `await Task.Delay`，我认为应该也不会让 UI 卡：

```CS
using CommunityToolkit.Mvvm.ComponentModel;
using System;
using System.Threading;
using System.Threading.Tasks;
namespace WpfApp1
{

internal partial class MainWindowViewModel: ObservableObject
{
    public MainWindowViewModel()
    {
        last = DateTime.Now;

        Task readAxiSatatusTask = Task.Factory.StartNew(async () =>

        {

            while (true)

            {

                await Task.Delay(100);

                if (Cnt < 100000)

                    Cnt += 1;

                else

                    Cnt = 0;

                Dt = (DateTime.Now - last).TotalMilliseconds;

                last = DateTime.Now;

            }

        }, TaskCreationOptions.LongRunning);

    }
    [ObservableProperty]
    private int cnt;
    private DateTime last;
    [ObservableProperty]
    private double dt=0;
}
}
```
